File: stock-market-service/src/main/java/rs/edu/raf/mapper/OrderMapper.java
package rs.edu.raf.mapper;

import lombok.AllArgsConstructor;
import org.springframework.stereotype.Component;
import rs.edu.raf.dto.*;
import rs.edu.raf.model.order.*;
import rs.edu.raf.security.JwtUtil;

@Component
@AllArgsConstructor
public class OrderMapper {
    private JwtUtil jwtUtil;

    public Order marketOrderCreateDTOtoOrder(MarketOrderCreateDTO marketOrderCreateDTO) {
        MarketOrder order = new MarketOrder();

        order.setQuantity(marketOrderCreateDTO.quantity());
        order.setTicker(marketOrderCreateDTO.ticker());
        order.setOrderAction(OrderAction.valueOf(marketOrderCreateDTO.orderAction()));
        order.setOrderCreator(jwtUtil.getOrderCreator());

        return order;
    }

    public Order limitOrderCreateDTOtoOrder(LimitOrderCreateDTO limitOrderCreateDTO) {
        LimitOrder order = new LimitOrder();

        order.setQuantity(limitOrderCreateDTO.quantity());
        order.setTicker(limitOrderCreateDTO.ticker());
        order.setOrderAction(OrderAction.valueOf(limitOrderCreateDTO.orderAction()));
        order.setLimitAmount(limitOrderCreateDTO.limit());
        order.setOrderCreator(jwtUtil.getOrderCreator());

        return order;
    }

    public Order stopOrderCreateDTOtoOrder(StopOrderCreateDTO stopOrderCreateDTO) {
        StopOrder order = new StopOrder();

        order.setQuantity(stopOrderCreateDTO.quantity());
        order.setTicker(stopOrderCreateDTO.ticker());
        order.setOrderAction(OrderAction.valueOf(stopOrderCreateDTO.orderAction()));
        order.setStop(stopOrderCreateDTO.stop());
        order.setOrderCreator(jwtUtil.getOrderCreator());

        return order;
    }

    public Order stopLimitOrderCreateDTOtoOrder(StopLimitOrderCreateDTO stopLimitOrderCreateDTO) {
        StopLimitOrder order = new StopLimitOrder();

        order.setQuantity(stopLimitOrderCreateDTO.quantity());
        order.setTicker(stopLimitOrderCreateDTO.ticker());
        order.setOrderAction(OrderAction.valueOf(stopLimitOrderCreateDTO.orderAction()));
        order.setStop(stopLimitOrderCreateDTO.stop());
        order.setLimitAmount(stopLimitOrderCreateDTO.limit());
        order.setOrderCreator(jwtUtil.getOrderCreator());

        return order;
    }

    public Order allOrNoneOrderCreateDTOtoOrder(AllOrNoneOrderCreateDTO allOrNoneOrderCreateDTO) {
        StopLimitOrder order = new StopLimitOrder();

        order.setQuantity(allOrNoneOrderCreateDTO.quantity());
        order.setTicker(allOrNoneOrderCreateDTO.ticker());
        order.setOrderAction(OrderAction.valueOf(allOrNoneOrderCreateDTO.orderAction()));
        order.setOrderCreator(jwtUtil.getOrderCreator());

        return order;
    }


    public OrderDTO orderToOrderDTO(Order order) {
        return new OrderDTO(order.getId(), order.getOrderAction().toString(), order.getQuantity(), order.getTicker());
    }
}

File: stock-market-service-kotlin/src/main/kotlin/org/example/stockmarketservicekotlin/mapper/OrderMapper.kt
package org.example.stockmarketservicekotlin.mapper

import org.example.stockmarketservicekotlin.model.order.*
import org.example.stockmarketservicekotlin.dto.*
import org.example.stockmarketservicekotlin.security.JwtUtil

import org.springframework.stereotype.Component

@Component
class OrderMapper(private val jwtUtil: JwtUtil) {
    fun marketOrderCreateDTOtoOrder(marketOrderCreateDTO: MarketOrderCreateDTO): Order {
        val orderAction = OrderAction.valueOf(marketOrderCreateDTO.orderAction)
        return MarkerOrder(
            null,
            orderAction,
            marketOrderCreateDTO.ticker,
            marketOrderCreateDTO.quantity,
            jwtUtil.getIDForLoggedUser()!!
        )
    }

    fun limitOrderCreateDTOtoOrder(limitOrderCreateDTO: LimitOrderCreateDTO): Order {
        val orderAction = OrderAction.valueOf(limitOrderCreateDTO.orderAction)
        return LimitOrder(
            null,
            orderAction,
            limitOrderCreateDTO.ticker,
            limitOrderCreateDTO.quantity,
            jwtUtil.getIDForLoggedUser()!!,
            limitOrderCreateDTO.limit
        )
    }

    fun stopOrderCreateDTOtoOrder(stopOrderCreateDTO: StopOrderCreateDTO): Order {
        val orderAction = OrderAction.valueOf(stopOrderCreateDTO.orderAction)
        return StopOrder(
            null,
            orderAction,
            stopOrderCreateDTO.ticker,
            stopOrderCreateDTO.quantity,
            jwtUtil.getIDForLoggedUser()!!,
            stopOrderCreateDTO.stop
        )
    }

    fun stopLimitOrderCreateDTOtoOrder(stopLimitOrderCreateDTO: StopLimitOrderCreateDTO): Order {
        val orderAction = OrderAction.valueOf(stopLimitOrderCreateDTO.orderAction)
        return StopLimitOrder(
            null,
            orderAction,
            stopLimitOrderCreateDTO.ticker,
            stopLimitOrderCreateDTO.quantity,
            jwtUtil.getIDForLoggedUser()!!,
            stopLimitOrderCreateDTO.stop,
            stopLimitOrderCreateDTO.limit
        )
    }

    fun allOrNoneOrderCreateDTOtoOrder(allOrNoneOrderCreateDTO: AllOrNoneOrderCreateDTO): Order {
        val orderAction = OrderAction.valueOf(allOrNoneOrderCreateDTO.orderAction)
        return AllOrNoneOrder(
            null,
            orderAction,
            allOrNoneOrderCreateDTO.ticker,
            allOrNoneOrderCreateDTO.quantity,
            jwtUtil.getIDForLoggedUser()!!
        )
    }


    fun orderToOrderDTO(order: Order): OrderDTO {
        return OrderDTO(order.id, order.orderAction.toString(), order.quantity, order.ticker)
    }
}